Basic tests

vpodzime opened the pull request:
and a couple fixes for discovered issues.

vpodzime committed:
ci: Add shfmt checking of src/gen_docker-compose

Ticket: none
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
fix: Fix appending to the device_types array

Appending to an array is done with `array+=(item)`, without
parantheses, it's just appended as a string to the first element
in the array.

Ticket: MEN-8795
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
test: Add a basic set of tests

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
chore: Add log message when there is nothing to rollback to

Ticket: MEN-8793
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
test: Add a test for second ArtifactInstall failing

Adding a test to make sure proper handling of a failure in
artifact installation on top of a committed artifact.

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
fix: Stop and backup previous composition first in install_artifact()

In case ArtifactInstall fails, ArtifactRollback is run. In order
to make the rollback possible, we need to make sure
ArtifactInstall doesn't fail before it prepares ground for the
potential future rollback. Such preparation thus has to happen
ASAP in `install_artifact()`.

Also adapt tests to this new behavior.

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
ci: Run tests in CI

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+cleanup_file_tree() {
+    rm -rf "${WORKDIR}/artifact"
+    rm -rf "${WORKDIR}/images"
+    rm -rf "${WORKDIR}/manifests"
```
Looking at the clean-up, we could have had the file tree in `${WORKDIR}/filetree/...` and clean that parent dir at once.

vpodzime replied:
The file tree is in `/artifact`, `/images` and `/manifests` are removed here to make sure some old images and manifests are not included in follow-up tests. Now that I see this, I think two changes would be nice:
- rename `/artifact` to `/artifact-file-tree`
- remove `/images` and `/manifests` as soon as they are not needed, i.e. when they are packed into the respective tarballs

vpodzime replied:
I changed my mind. The second change above makes things messy, it's better to the cleanup at one place. So I'll keep that part as-is.

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+
+cleanup_docker_mock() {
+    rm -rf "${WORKDIR}/bin"
+    truncate -s0 "$CMDLINE_LOGGER_LOG_FILE"
```
TIL `truncate` :raised_hands:

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+        return 1
+    fi
+    if [ -d "${PERSISTENT_DIR}/cleanup" ]; then
+        echo "'cleanup' directory exists after rollback"
```
This comment may indicate a bug: isn't it cleanup supposed to be called after commit and rollback? In other words I would expect cleanup to exist after rollback and be removed only after cleanup.

vpodzime replied:
There is a cleanup happening above. I updated the error message.

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+}
+tests+=(test_artifact_install_commit)
+
+test_artifact_install_rollback_cleanup() {
```
Related to previous comment, should we have a test for `test_artifact_install_rollback` (no cleanup)? There I would expect `new` and `current` to be removed but `cleanup` to still survive.

vpodzime replied:
Sure, I can add such a test.

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+}
+tests+=(test_artifact_install_rollback_cleanup)
+
+test_artifact_install_commit_install_commit_cleanup() {
```
This and the next test, the point is that they have an already installed artifact and then do the second one (the test). Consider making this explicit in the titles with something like
* `test_two_artifacts_install_commit_cleanup` / `test_two_artifacts_install_rollback_cleanup`
* `test_existing_artifact_install_commit_cleanup` / `test_existing_artifact_install_rollback_cleanup`
* or something like that

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+
+    cat << EOF | diff -u - "$CMDLINE_LOGGER_LOG_FILE" || rc=$?
+docker image load
+docker image load
```
By the way, these docker load _nothing_  I've been seeing and are hard to verify that are called correctly. In this test particularly as we load different images it would have value.

With [the suggestion for no stdin](https://github.com/mendersoftware/mender-container-modules/pull/2#discussion_r2410414102) this could be verified.

vpodzime replied:
Yup, I cannot disagree. I'll change it in a follow-up commit. (just to avoid complicating rebasing for me)

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+
+    # For this test, we need 'docker images...' to return something consistent
+    # between runs that we can easily refer to.
+    cat << EOF > "${WORKDIR}/bin/docker"
```
This mock could have been moved to a shared function between these two tests that required the previous artifact.

vpodzime replied:
True, but having it in the test is also nice because it's quite tightly related to what the test is doing. We may refactor this in the future, of course.

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+    cat << EOF > "${WORKDIR}/manifests/docker-compose.yml"
+services:
+  php:
+    image: bad/php:worst
```
:laughing:

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+EOF
+
+    "${SRCDIR}/docker-compose" ArtifactInstall "${WORKDIR}/artifact" > "${WORKDIR}/docker-compose.log" 2>&1 || rc=$?
+    if [ $rc -eq 0 ]; then
```
For consistency with the rest of the test suite, why not `-ne 1` ? It is cleaner to check for an specific expectation and treat the rest as errors.

lluiscampos replied:
(there are more iterations of this in upcoming commits - YMMV :rofl: )

vpodzime replied:
Yup. I didn't want to worry about specific error exit codes, but you're right.

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -0,0 +1,660 @@
+    exit 0
+else
+    exit 1
+fi
```
Really nice tests, I must admit! Well done :hamburger:

lluiscampos commented on tests/test_docker-compose.sh:
```diff
@@ -627,6 +627,244 @@ EOF
 }
 tests+=(test_artifact_install_load_fail_rollback_cleanup)
 
+test_artifact_install_up_fail_rollback_cleanup() {
```
Nit pick: this commit adds two tests: the mentioned in the commit and this one (it was confusing to get the first test the one ignored in the commit)

vpodzime replied:
That's a mistake, I meant to add the first one to the basic set. I'll fix it.

lluiscampos commented on .gitlab-ci.yml:
```diff
@@ -12,3 +12,11 @@ include:
     - SHELL_SCRIPTS="src/docker-compose src/gen_docker-compose tests/test_docker-compose.sh"
+
+test:unit-tests:
+  image: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/debian:12
```
Maybe make this 2 more years future proofed with `debian:13`?

vpodzime replied:
Sure, why not. :) :+1:

lluiscampos reviewed (approved):
Overall looks very good! The fix also makes sense to me.

I didn't find any blockers, follow-up as you see fit.

vpodzime committed:
chore: Use the `--input` option when loading images

The difference between `docker image load < /some/image` and
`docker image load --input /some/image` is that the latter can
handle additional image formats and URIs. We don't need this
functionality (and non-standard image formats and URIs should be
discouraged), but using the option allows us to better check the
arguments in our docker mock in scripts.

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
chore: Add tests-related Makefile targets

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime committed:
ci: Run CI on oslo-xps

It's cheap and fast and we can easily change this if there are
any issues with it in the future.

Ticket: MEN-8794
Changelog: none
Signed-off-by: Vratislav Podzimek <vratislav.podzimek+auto-signed@northern.tech>

vpodzime commented:
> I didn't find any blockers, follow-up as you see fit.

I followed up on almost everything. Tentatively merging this now, but feel free to suggest follow-ups, e.g. related to code coverage (it's over 70%, btw).
